clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -72,28 +78,73 @@ static uint64 instr_count;
 // but to avoid the complexity of different instrumentation we need a threshold.
 #define DELAY_FOREVER_THRESHOLD (1024 * 1024 * 1024)
 
-static std::atomic<bool> reached_trace_after_instrs;
+static std::atomic<bool> reached_trace_after_instrs(false);
 std::atomic<uint64> retrace_start_timestamp;
 
+static std::atomic<uint64> instr_intervals_idx(0);
+static std::vector<uint64> trace_for_instrs_list;
+static std::vector<uint64> no_trace_for_instrs_list;
+
+void
+increment_instr_intervals_index()
+{
+    if (!trace_for_instrs_list.empty())
+        instr_intervals_idx.fetch_add(1);
+}
+
+uint64
+get_trace_after_instrs_value()
+{
+    if (op_trace_after_instrs.get_value() > 0)
+        return op_trace_after_instrs.get_value();
+    if (!no_trace_for_instrs_list.empty())
+        return no_trace_for_instrs_list[0];
+    return 0;
+}
+
+uint64
+get_trace_for_instrs_value()
+{
+    if (op_trace_for_instrs.get_value() > 0)
+        return op_trace_for_instrs.get_value();
+    if (!trace_for_instrs_list.empty()) {
+        uint64 i = instr_intervals_idx.load(std::memory_order_acquire);
+        return trace_for_instrs_list[i];
+    }
+    return 0;
+}
+
+uint64
+get_retrace_every_instrs_value()
+{
+    if (op_retrace_every_instrs.get_value() > 0)
+        return op_retrace_every_instrs.get_value();
+    if (!no_trace_for_instrs_list.empty()) {
```

Comment:
It's less clear why no_trace_for_instrs_list should be checked in get_retrace_every_instrs_value. They seem logically different (even though for some usages handling both together may work). Maybe extract out the no_trace_for_instrs_list code into each caller where needed.

Reply:
They are related: `op_retrace_every_instrs` is the periodic "no trace period".
While `op_trace_after_instrs` is the initial "no trace period".
Maybe they should be renamed:
```
op_retrace_every_instrs -> op_wait_for_instrs
op_trace_after_instrs -> op_wait_for_instrs_initial
```
and
```
no_trace_for_instrs_list -> wait_for_instrs_list
```
?

Would that make it more clear? Although if we do that, should we rename the user facing op_ as well? Would that break things?

> Maybe extract out the no_trace_for_instrs_list code into each caller where needed

But I think every caller needs it, because I'm using the same window tracing mechanism we had before.
I built the irregular windows tracing on top of the 3 op_* we already had, I'm just updating their values (that used to be constant) depending on the window.

Reply:
Wouldn't be worth changing a user-facing option for this issue.

I agree that there's an overlap between the meaning of no_trace_for_instrs_list and op_retrace_every_instrs. But this function is meant to return the retrace_every_instrs value, and the former cannot be called the latter, even though the latter may be called the former in some sense.

Reply:
Is this just a naming issue or am I missing something?
I renamed it to: `get_current_no_trace_for_instrs_value()` following https://github.com/DynamoRIO/dynamorio/pull/6927#discussion_r1726159360